### ISSUE TITLE
Udt with Oracle.ManagedDataAccess.Core

### ISSUE THREAD

**Issue by preslyel** [oracle/dotnet-db-samples#56]
I can't find any information about UDT with Oracle.ManagedDataAccess.Core. Does package support UDT? And if it does, can you provide some sample?

**Comment by alexkeh** [comment]
UDTs with ODP.NET Core is not currently supported. We're working on porting the unmanaged code to the managed providers. It is significant work, which is why it is taking some time.

**Comment by caroncyp** [comment]
Do you know how much time you will require for UDT to be supported ?

**Comment by alexkeh** [comment]
I'll have a better idea on target dates late in the second half of this year.

**Comment by Victor-Tseng** [comment]
@alexkeh In the meantime do you have any suggestion if I want to migrate .NET applications that use UDT to .NET Core now? We are now evaluating the feasibility of containerizing existing .NET applications into Linux container and therefore we need to migrate to .NET Core first.

**Comment by alexkeh** [comment]
@Victor-Tseng Some customers are converting the UDT data to well-known text or XML. JSON is another consideration.

**Comment by jeroenvheel** [comment]
How are the udt’s coming along? I checked the oracle 19 driver unmanaged and it’s build on .net framework 4 and not .net standard 2.0

**Comment by alexkeh** [comment]
So far, so good.

The unmanaged driver will remain specific to .NET 4 and .NET 2/3.

**Comment by jeroenvheel** [comment]
Do you know in which version it’s going to land?

**Comment by alexkeh** [comment]
By the second half of the year, I'll have a much better idea.

**Comment by rsankati** [comment]
Will Unmanaged drive  supports .net core >=2?

**Comment by alexkeh** [comment]
No. Unmanaged ODP.NET has a dependency on platform specific unmanaged code. The plan is to port UDT functionality to ODP.NET Core.

**Comment by rsankati** [comment]
Thanks Alex but we are planning to migrate from .net framewrk to core where all udt types depedencies  are used in exiting .net  frame work. so what is alternative to use udttypes ?
will porting UDT funtionality to ODP.NET Core done in near future?
 and also *OracleDbType.Array* is not availble in Core dll is there any alternative to use ?

**Comment by alexkeh** [comment]
Some customers are using well-known text as an alternative to UDTs for now. I'll have a better idea on the release plans after the next Oracle DB release goes out. My current guess would be sometime next year.

UDTs are a sizable project, which is why it's taking so long. Moreover, it has a sizeable infrastructure underneath it that precludes releasing anything too piecemeal.

**Comment by chris5287** [comment]
Is support for BOOLEAN being added as well, or is there another issue to follow for this?

**Comment by alexkeh** [comment]
@chris5287 OracleBoolean is currently supported by managed and Core ODP.NET

https://docs.oracle.com/en/database/oracle/oracle-data-access-components/19.3/odpnt/OracleBooleanStructure.html#GUID-B5B5ACB7-14A5-43D5-B8A9-7D0046D40E2E

**Comment by chris5287** [comment]
> @chris5287 OracleBoolean is currently supported by managed and Core ODP.NET
> 
> 
> 
> https://docs.oracle.com/en/database/oracle/oracle-data-access-components/19.3/odpnt/OracleBooleanStructure.html#GUID-B5B5ACB7-14A5-43D5-B8A9-7D0046D40E2E

I didn’t know that existed, assume this article needs updating then: https://docs.oracle.com/cd/B19306_01/win.102/b14306/appendixa.htm#BEHCHBJF

**Comment by alexkeh** [comment]
@chris5287 That "article" is the documentation for Oracle DB Extensions for .NET version 10.2. It is accurate for the 10.2 version, which did not support Booleans.

I linked to the ODP.NET 19c version. Managed ODP.NET supported Booleans since 12.2 and ODP.NET Core since 18.3.

**Comment by chris5287** [comment]
Thanks @alexkeh for the details, sadly the server we are connecting to currently is 11.2 :(

**Comment by alexkeh** [comment]
@chris5287 PL/SQL Booleans are supported in DB 11.2 and are certified to interoperate with ODP.NET 18c and 19c. You should be able to use PL/SQL Booleans today as long as you use a more recent ODP.NET version.

**Comment by chris5287** [comment]
Which version of `Oracle.ManagedDataAccess.Core` should work, I've just tried `2.19.70` but i still get the `ORA-03115: unsupported network datatype or representation` error for OracleBoolean values

**Comment by alexkeh** [comment]
I put together a simple stored procedure:

```
PROCEDURE "TESTBOOL" (
  "PARAM1" IN BOOLEAN, 
  "PARAM2" OUT BOOLEAN) IS
BEGIN
	PARAM2 := PARAM1;
END;
```
And a C# console app:
```
            string constr = "<Connection String>";
            OracleConnection con = new OracleConnection(constr);
            con.Open();

            OracleCommand cmd = new OracleCommand("TESTBOOL", con);
            cmd.CommandType = CommandType.StoredProcedure;

            OracleParameter oparam = cmd.Parameters.Add("PARAM1", OracleDbType.Boolean);
            oparam.Direction = ParameterDirection.Input;
            oparam.Value = "True";

            OracleParameter oparam2 = cmd.Parameters.Add("PARAM2", OracleDbType.Boolean);
            oparam2.Direction = ParameterDirection.Output;

            cmd.ExecuteNonQuery();

            Console.WriteLine(oparam2.Value.ToString());
```
Does this sample work in your environment?

**Comment by chris5287** [comment]
Sadly not that simple as it’s a vendor database (EBS)

**Comment by chris5287** [comment]
FYI we are trying to call EBS APIs, so it’s a package procedure call, not just a stored procedure, if that helps understand the error we receive

**Comment by alexkeh** [comment]
Calling a package is not that different. You would reference the package and stored procedure as;
`OracleCommand cmd = new OracleCommand("Package.Procedure", con);`

**Comment by tinyg** [comment]
Hi, Just wondering if there's any update on when this feature will be ready? This driver is now the only remaining dependency that's blocking us from upgrading to .NET Core.

The WKT works barely ok (2x slower in our application) for reading but not well at all for writing - SQL needs major rewrites in some cases.

**Comment by alexkeh** [comment]
I don't have an update yet. We're working on delivering something next year, but we'll need some more time before we can provide a more specific time frame.

**Comment by omora** [comment]
Please our PL SQL code is full of UDTs. 
that's blocking us from upgrading to .NET Core.

At the moment we are testing other drivers (devart) but we will like to continue using Oracle!!

**Comment by alexkeh** [comment]
I'll provide an expected delivery time frame as soon as possible.

**Comment by alexkeh** [comment]
The dev team is making good progress on UDT support for managed ODP.NET and ODP.NET Core. My expectation is that we'll deliver the feature in the first half of 2021.

**Comment by tinyg** [comment]
Thanks for letting us know Alex!

**Comment by Beej126** [comment]
with all respect to Alex (he's been a great stream of real info) - i'm compelled to share an alternative that looks like it's really going to save my bacon on a tight project timeline (when are they not?)... if anyone else out there needs this immediately and has $300 to spare, DevArt's Oracle provider makes elegant quick work of passing lists of objects to stored procs... it really works... fully .net core 3.1 compatible, even on linux, with no dependencies on native oracle client ... see working console app sample below based on a quick blurb i saw in their forums (check links in my comments below)... as you can see their "OracleType.GetObjectType()" API makes the UDT marshalling part of this incredibly easy for us... way better than even in the old unmanaged ODP ... i plan to leave a lot of my code tied to the official oracle provider and only sprinkle in this new dependency where i need table type support.

```csharp
using System;
using System.Data;
using Devart.Data.Oracle;
namespace ConsoleApp1
{
    class Program
    {
        private static int oraTable;

        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");

            //good docs:
            //direct connection: https://www.devart.com/dotconnect/oracle/docs/StoredProcedures-OracleCommand.html
            //linux licensing: https://www.devart.com/dotconnect/oracle/docs/?LicensingStandard.html
            using OracleConnection db = new OracleConnection("{insert yours}");

            //devart trial licensing nutshell... on WINDOWS, download & run their installer...
            //the only thing you really need from that install is the license key file dropped here: 
            //   %programdata%\DevArt\License\Devart.Data.Oracle.key
            // then just go nuget the "Devart.Data.Oracle" package reference
            //on Windows, the trial license file gets automatically picked up by their runtime
            //if you're on Linux, basically just read their good instructions out there
            //i've just tested it on WSL so far and plan to try on Azure linux app svc within next day
            //db.ConnectionString = "license key=trial:Devart.Data.Oracle.key;" + db.ConnectionString;

            db.Direct = true; //nugget: crucial!! https://www.devart.com/dotconnect/oracle/docs/DirectMode.html
            db.Open();

            var cmd = db.CreateCommand("UserPermissions_u", CommandType.StoredProcedure);
            cmd.DeriveParameters();
            //tblParm.OracleDbType = OracleDbType.Table;

            //passing "table" type proc parm example: https://forums.devart.com/viewtopic.php?t=22243
            var obj = new OracleObject(OracleType.GetObjectType("UserPerm", db));
            var tbl = new OracleTable(OracleType.GetObjectType("UserPerms", db));

            obj["UserPermissionId"] = "sR1CKjKYSKvgU90GUgqq+w==";
            obj["adv"] = 1;
            tbl.Add(obj);
            cmd.Parameters["IN_Email"].Value = "banderson@kingcounty.gov";
            cmd.Parameters["IN_Permissions"].Value = tbl;
            cmd.ExecuteNonQuery();

            //"i can't believe it's not butter!" -me, just now =)
        }
    }
}
```

here's the corresponding oracle side definitions:
```sql
create or replace type UserPerm as object ( UserPermissionId varchar2(24), std number(1), adv number(1)  );
create or replace type UserPerms as table of SeedPak.UserPerm;

create or replace PROCEDURE UserPermissions_u (
  IN_Email IN varchar2,
  IN_Permissions IN UserPerms
) is 

dummyvar number default 0;

begin

select count(*) into dummyvar from table(IN_Permissions);

end;
/

```

good luck and peace to all

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
https://www.nuget.org/packages/Oracle.ManagedDataAccess/

https://www.nuget.org/packages/Oracle.ManagedDataAccess.Core

We are using the .Net Framework version (Not .Core).   UDT appears to be limited in that version.   Will it be available in both nuget packages?

**Comment by alexkeh** [comment]
@s3YwCf2ZbfJG4SHAfjQMAjtsf 
Yes, UDTs are planned to be in both packages sometime in the first half of 2021.

**Comment by Assam** [comment]
Are we there yet? If not can we please get a preview version to test the existing code?

**Comment by alexkeh** [comment]
@Assam It's very close. Around the end of February, we'll have a validation drop we can provide supported customers. If you'd like access, email us dotnet_us(at)oracle.com.

**Comment by xtnbct** [comment]
Hi! 
Read about this and wondering if we can also get a preview version of managed dll for .NET Framework that supports UDTs?
If possible to participate in beta testing as well?

Thanks!

**Comment by alexkeh** [comment]
@xtnbct 
Yes, send us an email dotnet_us(at)oracle.com for us to arrange access for you. This validation drop is not available to the general public. So, it won't be available to download publicly until we go production.

**Comment by sumitgarg09** [comment]
Hi Alex,

Any good news, when we can use UDTs with general public!

**Comment by alexkeh** [comment]
@sumitgarg09 
The validation drop testing just started. So no major news to share at this point. We'll see how the user experience is.

**Comment by JFMarty** [comment]
Hi, 
any updates how the validation is going and when we can expect the public version?

**Comment by alexkeh** [comment]
Our target release date is end of this month. I can't promise it will exactly be then, especially as that day falls on a US holiday, but somewhere thereabouts.

**Comment by alexkeh** [comment]
End of May looks very unlikely now for the UDT production release. My new estimate is by end of June, back to the previous estimate.

**Comment by supermun** [comment]
@alexkeh please update information...

**Comment by alexkeh** [comment]
Our final pre-release validation drop to customers has been successful so far. No issues found.

We're lining up the ODP.NET UDT 21.3 release with the main DB 21c on-premises release, which is set for mid-July.

**Comment by alexkeh** [comment]
Given that it's certain our next managed ODP.NET and ODP,NET Core 21.3 versions will deliver UDT support, I'm closing this issue.

The ODP.NET 21.3 release on NuGet Gallery is beholden to the Oracle DB 21c on-premises RTM, which introduces a variable outside my team's direct control. Nonetheless, I know the DB team is working on releasing 21c ASAP.

**Comment by Beej126** [comment]
Hi @alexkeh, i'm seeing [Oracle.ManagedDataAccess.Core v3.21.3](https://www.nuget.org/packages/Oracle.ManagedDataAccess.Core/3.21.3) released earlier this week... should we expect this to include the new UDT support? any chance there's a code sample somewhere demonstrating basic usage?

**Comment by alexkeh** [comment]
Yes, ODP.NET Core 21.3 includes UDT support. More details about the [new UDT feature, differences with unmanaged ODP.NET, and how to migrate to managed/Core ODP.NET UDT are in my blog post](https://medium.com/oracledevs/announcing-odp-net-21-3-user-defined-types-for-managed-odp-net-and-odp-net-core-836fa81da3ba).

[UDT samples](https://github.com/oracle/dotnet-db-samples/tree/master/samples/udt) are available on GitHub.

**Comment by Beej126** [comment]
> Yes, ODP.NET Core 21.3 includes UDT support... 
> [UDT samples](https://github.com/oracle/dotnet-db-samples/tree/master/samples/udt) are available on GitHub.

want to say i like the looks of this ... the [Nested-Table.cs](https://github.com/oracle/dotnet-db-samples/blob/master/samples/udt/Nested-Table.cs) sample is the most similar to my interest in passing object collections as stored proc parms... i like that you provide easy automatic marshalling from C# array of class to an oracle object array UDT with the same shape via direct value assignment... i know it took a lot of work to get to this point... thank you to you and your team!

**Comment by omora** [comment]
Just trying your example using ODP.NET Core 21.3 [Object-UDT.cs](https://github.com/oracle/dotnet-db-samples/blob/master/samples/udt/Object-UDT.cs)

I can not get it working ... ()

 System.InvalidOperationException : A01ARQ01.ODPNET_PERSON_TYPE not valid

  Seguimiento de la pila: 
    UDTNamedType.GetNamedTypeMetaData(OracleConnection conn, OracleCommand getTypeCmd)
    UDTTypeCache.GetMetaData(OracleConnection conn, UDTNamedType udtType)
    UDTTypeCache.CreateUDTType(OracleConnection conn, String schemaName, String typeName)
    UDTTypeCache.GetUDTType(OracleConnection conn, String schemaName, String typeName)
    OraclePoolManager.GetUDTType(OracleConnection conn, String schemaName, String typeName)
    OracleConnection.GetUDTTypeFromCache(String schemaName, String typeName)
    OracleParameter.GetUDTType(OracleConnection conn)
    OracleParameter.PreBind_UDT(OracleConnection conn)
    OracleParameter.PreBind(OracleConnectionImpl connImpl, ColumnDescribeInfo cachedParamMetadata, Boolean& bMetadataModified, Int32 arrayBindCount, ColumnDescribeInfo& paramMetaData, Object& paramValue, Boolean isEFSelectStatement, SqlStatementType stmtType)
    OracleCommandImpl.InitializeParamInfo(ICollection paramColl, OracleConnectionImpl connectionImpl, ColumnDescribeInfo[] cachedParamMetadata, Boolean& bMetadataModified, Boolean isEFSelectStatement, MarshalBindParameterValueHelper& marshalBindValuesHelper)
    OracleCommandImpl.ProcessParameters(OracleParameterCollection paramColl, OracleConnectionImpl connectionImpl, ColumnDescribeInfo[] cachedParamMetadata, Boolean& bBindMetadataModified, Boolean isEFSelectStatement, MarshalBindParameterValueHelper& marshalBindValuesHelper)
    OracleCommandImpl.ExecuteNonQuery(String commandText, OracleParameterCollection paramColl, CommandType commandType, OracleConnectionImpl connectionImpl, Int32 longFetchSize, Int64 clientInitialLOBFS, OracleDependencyImpl orclDependencyImpl, Int64[]& scnFromExecution, OracleParameterCollection& bindByPositionParamColl, Boolean& bBindParamPresent, OracleException& exceptionForArrayBindDML, OracleConnection connection, OracleLogicalTransaction& oracleLogicalTransaction, Boolean isFromEF)
    OracleCommand.ExecuteNonQuery()
    DbCommand.ExecuteNonQueryAsync(CancellationToken cancellationToken)

Oracle setup
`CREATE OR REPLACE TYPE ODPNET_PERSON_TYPE AS OBJECT
    (NAME VARCHAR2(30), ADDRESS VARCHAR2(60), AGE NUMBER(3))`
.net implementation
Person is the some as your example

`        public async Task<bool> Insert_Person(Person persona)
        {
            bool ret = false;
            using var cmd = new OracleCommand("A01ARQ01.ODPNET_INSERT_CONTACTS")
            {
                CommandType = System.Data.CommandType.StoredProcedure
                
            };
            
            OracleParameter param1 = new OracleParameter();
            param1.ParameterName = "PARAM1";
            param1.OracleDbType = OracleDbType.Object;
            param1.Direction = ParameterDirection.Input;

            // Note: The UdtTypeName is case-senstive
            param1.UdtTypeName = "A01ARQ01.ODPNET_PERSON_TYPE";
            param1.Value = persona;


            cmd.Parameters.Add(param1);
            

            var numrows = await ExecuteNonQueryAsync(cmd);
            if (numrows > 0)
                ret = true;
            return ret;
        }`

**Comment by leojth** [comment]
> 
> 
> Yes, ODP.NET Core 21.3 includes UDT support. More details about the [new UDT feature, differences with unmanaged ODP.NET, and how to migrate to managed/Core ODP.NET UDT are in my blog post](https://medium.com/oracledevs/announcing-odp-net-21-3-user-defined-types-for-managed-odp-net-and-odp-net-core-836fa81da3ba).

Just converted a project using SDO Geometry to the new managed driver. It only took a couple of minutes to convert and it works like a charm. Great job!

**Comment by Beej126** [comment]
> Just trying your example using ODP.NET Core 21.3 [Object-UDT.cs](https://github.com/oracle/dotnet-db-samples/blob/master/samples/udt/Object-UDT.cs)
> I can not get it working ... 

just taking a guess from a glance... try removing the owner/schema prefix on the UDT in C#, the reference sample didn't have that.
```
param1.UdtTypeName = "**A01ARQ01**.ODPNET_PERSON_TYPE";
```

also double check the usage of what you've declared on the oracle side vs referencing on C# side... your last error message says "ODP_OBJ1_SAMPLE_PERSON_TYPE" but i don't see that in your oracle setup.

**Comment by alexkeh** [comment]
@omora I think the problem is the same one @Beej126 identified.

> also double check the usage of what you've declared on the oracle side vs referencing on C# side... your last error message says "ODP_OBJ1_SAMPLE_PERSON_TYPE" but i don't see that in your oracle setup.

Either you have to create the "ODP_OBJ1_SAMPLE_PERSON_TYPE" in the DB or remove references to the object in your code.

**Comment by omora** [comment]
Thanks for your support. i was pointing to a wrong Oracle environment.

**Comment by vvsibirkrv** [comment]
hello! 
we tried use UDT-type from Oracle.ManagedDataAccess 21.3 (array of numbers realized as UDT-type) in stored proc and get that error:
Procedure: 
ORA-06550: Строка 1, столбец 13:
PLS-00306: ошибочно число или типы аргументов при обращении к 'GET_TYPE_SHAPE'
ORA-06550: Строка 1, столбец 7:
PL/SQL: Statement ignored

DB version we have is 11.2.0.4.0. How can fix that error?

**Comment by alexkeh** [comment]
@vvsibirkrv It's hard to tell without knowing more about your code or what you changed to migrate to manage ODP.NET. See my blog post on the migration steps needed.

**Comment by vvsibirkrv** [comment]
@alexkeh since we are using the "table of number" type, I have not found another example like [Nested-Table.cs](https://github.com/oracle/dotnet-db-samples/blob/master/samples/udt/Nested-Table.cs)
I tried to run this code on our configuration (DB - 11.2.0.4.0, and ManagedOracleDataAccess 21.3) - got the same error:
ORA-06550: line 1, column 13:
PLS-00306: wrong number or types of arguments when accessing 'GET_TYPE_SHAPE'
ORA-06550: line 1, column 7:
PL / SQL: Statement ignored

Maybe something needs to be tweaked in the database?

**Comment by alexkeh** [comment]
@vvsibirkrv You shouldn't need to change anything in the DB if this UDT worked before with unmanaged ODP.NET. The changes only need to occur on the client side.

Look over the [ODP.NET UDT differences between managed and unmanaged and migration recommendations](https://medium.com/oracledevs/announcing-odp-net-21-3-user-defined-types-for-managed-odp-net-and-odp-net-core-836fa81da3ba). That should guide you what .NET code needs to be modified to migrate.

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
@alexkeh I was able to convert the project using the unmanaged client to managed and compile it.

I am trying work through one of the use cases that passes a parameter to a proc that is a UDT.

Oracle reports an error **{User}.{Type} is invalid**.  However, when I am building the OracleParameter, I am specifying UdtTypeName as just "Type".   However, I do see the type in the db.   

The error suggests that the schema is determined along the stack trace.   It seems like I have to specify the schema at some point.   From what I have read online, "ALTER SESSION SET CURRENT_SCHEMA=animals".   However, I have 1 OracleCommand that is already defined for the stored proc with parameters in a using statement.  

   at OracleInternal.UDT.Types.UDTNamedType.GetNamedTypeMetaData(OracleConnection conn, OracleCommand getTypeCmd)
   at OracleInternal.UDT.Types.UDTTypeCache.GetMetaData(OracleConnection conn, UDTNamedType udtType)
   at OracleInternal.UDT.Types.UDTTypeCache.CreateUDTType(OracleConnection conn, String schemaName, String typeName)
   at OracleInternal.UDT.Types.UDTTypeCache.GetUDTType(OracleConnection conn, String schemaName, String typeName)
   at OracleInternal.ConnectionPool.OraclePoolManager.GetUDTType(OracleConnection conn, String schemaName, String typeName)
   **at Oracle.ManagedDataAccess.Client.OracleConnection.GetUDTTypeFromCache(String schemaName, String typeName)**
   at Oracle.ManagedDataAccess.Client.OracleParameter.GetUDTType(OracleConnection conn)
   at Oracle.ManagedDataAccess.Client.OracleParameter.PreBind_UDT(OracleConnection conn)
   at Oracle.ManagedDataAccess.Client.OracleParameter.PreBind(OracleConnectionImpl connImpl, ColumnDescribeInfo cachedParamMetadata, Boolean& bMetadataModified, Int32 arrayBindCount, ColumnDescribeInfo& paramMetaData, Object& paramValue, Boolean isEFSelectStatement, SqlStatementType stmtType)
   at OracleInternal.ServiceObjects.OracleCommandImpl.InitializeParamInfo(ICollection paramColl, OracleConnectionImpl connectionImpl, ColumnDescribeInfo[] cachedParamMetadata, Boolean& bMetadataModified, Boolean isEFSelectStatement, MarshalBindParameterValueHelper& marshalBindValuesHelper)
   at OracleInternal.ServiceObjects.OracleCommandImpl.ProcessParameters(OracleParameterCollection paramColl, OracleConnectionImpl connectionImpl, ColumnDescribeInfo[] cachedParamMetadata, Boolean& bBindMetadataModified, Boolean isEFSelectStatement, MarshalBindParameterValueHelper& marshalBindValuesHelper)
   at OracleInternal.ServiceObjects.OracleCommandImpl.ExecuteNonQuery(String commandText, OracleParameterCollection paramColl, CommandType commandType, OracleConnectionImpl connectionImpl, Int32 longFetchSize, Int64 clientInitialLOBFS, OracleDependencyImpl orclDependencyImpl, Int64[]& scnFromExecution, OracleParameterCollection& bindByPositionParamColl, Boolean& bBindParamPresent, OracleException& exceptionForArrayBindDML, OracleConnection connection, OracleLogicalTransaction& oracleLogicalTransaction, Boolean isFromEF)
   at Oracle.ManagedDataAccess.Client.OracleCommand.ExecuteNonQuery()
   at CI.ConnectHsDbAccess.HSDbHandlers.HSDbConnection.ExecStoredProcedure(String procedure_name, HSDbParameterList parameters) in C:\PJT\CI.HSNugetPackages\3.HelperLibraries\CI.ConnectHsDbAccess\HSDbHandlers\HsDBConnection.cs:line 869

**Comment by alexkeh** [comment]
@s3YwCf2ZbfJG4SHAfjQMAjtsf Does it work if you specify the schema the UDT is defined in?

**Comment by vvsibirkrv** [comment]
@alexkeh i run [Nested-Table.cs](https://github.com/oracle/dotnet-db-samples/blob/master/samples/udt/Nested-Table.cs) - from your examples. I found [article](http://ksun-oracle.blogspot.com/2017/05/jdbc-oracle-objectcollection.html), where it say, about  **DBMS_PICKLER.GET_TYPE_SHAPE** -  "_whereas 12cR1 APIs are enhanced with two more overloaded functions having 9 parameters_"
maybe need is some patch?

**Comment by alexkeh** [comment]
@vvsibirkrv Ah, yes! I forgot that the new UDT APIs require Oracle DB 12.1 or higher. You will need to upgrade your DB from 11.2.0.4.

**Comment by Beej126** [comment]
ouch, i was wondering if that shoe was going to drop. @alexkeh, i do appreciate you making that a clear requirement so quickly so we're not left wondering.

my shop is on 11.2 as well and it'll be at least a couple years before we could even consider upgrading because we're right in the middle of the project to build the new front end to replace the Oracle Forms app that is our main dependency on Oracle DB 11.2... we're choosing to invest our efforts in a new web UI with much more end user potential vs continuing to invest in Oracle Forms... removing that dependency also puts us in a better position to migrate the database to another platform entirely.

if anybody else is in this legacy Oracle situation, [DevArt's nuget packages that i mentioned previously in this thread](#issuecomment-714055092) have been working 100% solid for us on Oracle 11.2 supporting UDT's on .net core (including .net 5.0).

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
> @s3YwCf2ZbfJG4SHAfjQMAjtsf Does it work if you specify the schema the UDT is defined in?

How would I do that?    From what I read, the only way to specify the schema was by altering the session.   **ALTER SESSION SET CURRENT_SCHEMA=animals**  Is there anyway to control which Schema is used when calling a stored procedure on a single connection/command?

The connection string looks like this:

USER ID=**Schema_1**;POOLING=True;PASSWORD=**APWD**;DATA SOURCE="(DESCRIPTION = (ADDRESS = (PROTOCOL = TCP)(HOST = The_Machine)(PORT = 1521))(CONNECT_DATA = (SERVER = DEDICATED)(SERVICE_NAME = sn)))"

If I create a user with an exact name of the schema, then it works.

**Comment by alexkeh** [comment]
@s3YwCf2ZbfJG4SHAfjQMAjtsf You can append the schema to the type references. For example if the type is defined in OTHERSCHEMA, you would use the following format:

`[OracleCustomTypeMapping("OTHERSCHEMA.MYTYPE")]`

You shouldn't need to use ALTER SESSION. Just make sure that the ODP.NET user has privileges to access the OTHERSCHEMA's UDT definition.

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
@alexkeh I tried doing that manually for now and it did not resolve the error.   

If that solution had worked, I would have had to assign the OracleCustomTypeMapping at runtime, as the database schema name for this application is not fixed.

**Comment by alexkeh** [comment]
@s3YwCf2ZbfJG4SHAfjQMAjtsf Most likely, there is a configuration issue with in the DB or your code. Verify that your DB user has permissions to access the UDT type in the other schema. You can do this purely from the DB without your .NET app. If the access isn't available, grant the necessary permissions.

Next, make sure you are referencing the UDT with correct schema references in your ODP.NET code.

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
@alexkeh I have confirmed that the user has privileges to these types.

However, the error is saying User.Type is invalid.   Shouldn't it be saying Schema.Type is invalid?   

Does the user name have to match the schema name?

**Comment by alexkeh** [comment]
@s3YwCf2ZbfJG4SHAfjQMAjtsf Yes. In Oracle DBs, User and Schema are effectively synonymous.

This issue is very difficult to debug without the code and the schema setup scripts. The fundamental problem is that ODP.NET can't find the UDT definition. The problem could be due to a code issue. It could be due to a DB issue. It could be due to a permission issue. It could be due to an ODP.NET bug, though this is unlikely considering unmanaged ODP.NET has the same error and UDTs have been widely used with it for well over a decade.

In terms of next step, you could open an SR with Oracle Support to debug the problem if your organization is reluctant to share the schema scripts and code publicly on GitHub.

**Comment by s3YwCf2ZbfJG4SHAfjQMAjtsf** [comment]
@alexkeh Thanks for your help.  I can specific the schema name on the udtTypeName of the OracleParameter.  

Among the other nuisances of this DB, I am now able to see results and confirm that the UDT parameters are passing around.

**Comment by alexkeh** [comment]
Locking thread as the UDT feature request has been resolved. If there is an issue with UDT ODP.NET functionality, please create a new issue.
